clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -633,37 +634,127 @@ static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
+                               "invalid SYCL Table file.");
+
+    auto ImageOrErr = readFile(Elems[0]);
+    if (!ImageOrErr)
+      return ImageOrErr.takeError();
+
+    auto PropertiesOrErr = readPropertyRegistryFromFile(Elems[1]);
+    if (!PropertiesOrErr)
+      return PropertiesOrErr.takeError();
+
+    auto SymbolsOrErr = readFile(Elems[2]);
+    if (!SymbolsOrErr)
+      return SymbolsOrErr.takeError();
+
+    SYCLImage Image;
+    Image.Image = std::move(*ImageOrErr);
+    Image.PropertyRegistry = std::move(**PropertiesOrErr);
+    Image.Entries = std::move(*SymbolsOrErr);
+    Image.CompileOptions = CompileOptions;
+    Image.LinkOptions = LinkOptions;
+    Images.push_back(std::move(Image));
+  }
+
+  return Images;
+}
+
+// Run clang-offload-wrapper's functionality
+// TODO: this function should be removed in favor of community flow
+Expected<StringRef> wrapSYCLBinariesFromFile(StringRef InputFile,
+                                             const ArgList &Args) {
+  auto OutputFileOrErr = createOutputFile(
+      sys::path::filename(ExecutableName) + ".sycl.image.wrapper", "bc");
+  if (!OutputFileOrErr)
+    return OutputFileOrErr.takeError();
+
+  StringRef OutputFilePath = *OutputFileOrErr;
+  if (Verbose || DryRun) {
+    errs() << formatv(" offload-wrapper: input: {0}, output: {1}\n", InputFile,
+                      OutputFilePath);
+    if (DryRun)
+      return OutputFilePath;
+  }
+
+  auto ImagesOrErr = readSYCLImagesFromTable(InputFile, Args);
+  if (!ImagesOrErr)
+    return ImagesOrErr.takeError();
+
+  auto &Images = *ImagesOrErr;
 
-  SmallVector<StringRef, 8> CmdArgs;
-  CmdArgs.push_back(*ClangOffloadWrapperPath);
-  CmdArgs.push_back(Saver.save("-o=" + *TempFileOrErr));
-  llvm::Triple HostTriple(
-      Args.getLastArgValue(OPT_host_triple_EQ, sys::getDefaultTargetTriple()));
-  CmdArgs.push_back(Saver.save("-host=" + HostTriple.str()));
   const llvm::Triple Triple(Args.getLastArgValue(OPT_triple_EQ));
-  SmallString<128> TargetTripleOpt = Triple.getArchName();
-  CmdArgs.push_back(Saver.save("-target=" + TargetTripleOpt));
-  CmdArgs.push_back("-kind=sycl");
-  CmdArgs.push_back("-batch");
-  CmdArgs.push_back(InputFile);
-  if (Error Err = executeCommands(*ClangOffloadWrapperPath, CmdArgs))
-    return std::move(Err);
-  return *TempFileOrErr;
+  LLVMContext C;
+  const StringRef ModuleName = "offload.wrapper.object";
+  Module M(ModuleName, C);
+  M.setTargetTriple(
+      Args.getLastArgValue(OPT_host_triple_EQ, sys::getDefaultTargetTriple()));
+  if (auto E = wrapSYCLBinaries(M, Images))
+    return E;
+
+  if (Args.hasArg(OPT_print_wrapped_module))
+    errs() << M;
+
+  // TODO: Once "llc tool->runCompile" migration is finished we need to remove
+  // this scope and use community flow.
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(OutputFilePath, FD))
+    return errorCodeToError(EC);
+
+  raw_fd_ostream OS(FD, true);
+  WriteBitcodeToFile(M, OS);
+  return OutputFilePath;
```

Comment:
This seems like a dead code, because `wrapSYCLBinaries` returns `Error::succcess()` unless something went wrong.
I assume that `Error::success()` is convertible to `true` and therefore we take early exit on line 744 for most of the cases. Am I missing something?

Reply:
`Error::success()` evaluates to false according to the doc. https://llvm.org/doxygen/classllvm_1_1Error.html#a981b4992a3b7cce718c7995a7d6193a0

Reply:
Oh, I always forget that LLVM/clang is known for `true-on-error` approach